utils/events.py

```diff
@@ -135,3 +153,305 @@ def ensure_event_happens(self, target_type, target_id, event_type):
+
+        # checking only common attributes
+        if 'target_name' in self.event_attrs and 'target_id' not in self.event_attrs:
+            try:
+                target_id = self._tool.process_id(self.event_attrs['target_type'].value,
+                                                  self.event_attrs['target_name'].value)
+                self.event_attrs['target_id'] = EventAttr(**{'target_id': target_id})
+            except ValueError:
+                # vm or host name isn't added to db yet. need to wait
+                return False
+
+        common_attrs = set(self.event_attrs).intersection(set(evt.event_attrs))
+        for attr in common_attrs:
+            if not self.event_attrs[attr].match(evt.event_attrs[attr]):
+                return False
+        else:
+            return True
+
+    def add_attrs(self, *attrs):
+        """
+        event consists of attributes like event_type, etc.
+        this method allows to add an attribute to event
+        """
+        if isinstance(attrs, Iterable):
+            for attr in attrs:
+                if attr.name == 'target_name':
+                    # this is artificial attr which will be converted to target_id during matching
+                    self.event_attrs[attr.name] = attr
+                elif attr.name in self._default_attrs:
+                    # type check was removed because sqlalchemy's python_type
+                    # and type of returned values are different
+                    self.event_attrs[attr.name] = attr
+                else:
+                    logger.warning('The attribute {} type {} is absent in DB '
+                                   'or type mismatch.'.format(attr.name, attr.type))
+        else:
+            raise ValueError("incorrect parameters are passed {}".format(attrs))
+        return self
+
+    def build_from_raw_event(self, evt):
+        """
+        helper method which takes raw event from event_streams and prepares event object
+        """
+        # checking is this param - raw event, populating fields by this data then
+        if self._is_raw_event(evt):
+            self._parse_raw_event(evt)
+        return self
+
+
+class EventListener(Thread):
+    """
+     accepts "expected" events, listens to db events and compares showed up events with expected
+     events. Runs callback function if expected events have it.
+    """
+    def __init__(self, appliance):
+        super(EventListener, self).__init__()
+
+        self._tool = EventTool(appliance)
+
+        self._events_to_listen = []
+        # last_id is used to ignore already arrived messages the database
+        # When database is "cleared" the id of the last event is placed here. That is then used
+        # in queries to prevent events of this id and earlier to get in.
+        self._last_processed_id = None
+        self._stop_event = ThreadEvent()
+
+    def set_last_record(self, evt=None):
+        if evt is not None:
```

Comment:
Yeh, sorry, that's the problem sometimes in review you miss the context of the file